### PR TITLE
'deltaY': preliminary feature-complete implementation

### PR DESCRIPTION
**Pull request by denim2x**
upcoming incremental commits

### CODE REVIEW

**Comment by asturur** [comment]
Hello cristian,
would be better if you comment my open pr about the deltay changes. so we can merge that and you can continue with the changes on top of that.

Do not mix with master yet, or later will be harder to merge the branches for me.

File: src/shapes/text.class.js [hunk 1]
```diff
@@ -1152,22 +1156,42 @@
...
+     * @returns {Object} thisArg
+     * @chainable
      */
     _setScript: function(line, char, schema) {
+      if (!this._checkLocation(line, char) || typeof schema !== 'object') {
+        return this;
+      }
+      if (typeof schema.size !== 'number' || typeof schema.baseline !== 'number') {
+        return this;
+      }
```
Comment (asturur) [inline src/shapes/text.class.js]: We do not take care of this kind of checks.
If the dev provide a line/char that does not exist, he will set the style for that numbers.
There may be reasons to do that.

File: src/shapes/text.class.js [hunk 2]
```diff
@@ -927,9 +931,9 @@
       var box = {
         width: width,
         left: 0,
-        height: style.fontSize,
+        height: style.fontSize + Math.abs(style.deltaY),
```
Comment (asturur) [inline src/shapes/text.class.js]: the height of the box is the fontSize. We should not move from this assumptions.

**Comment by asturur** [comment]
@denim2x i do not understand the purposes of this changes.
On what part of the feature are you working on?

File: src/shapes/text.class.js [hunk 3]
```diff
@@ -175,8 +175,8 @@
      * @default
      */
     superscript: {
-      size:      0.60, // fontSize factor
-      baseline:  -0.67  // baseline-shift factor (upwards)
+      size:      0.83, // fontSize factor
+      baseline: -0.36  // baseline-shift factor (upwards)
```
Comment (asturur) [inline src/shapes/text.class.js]: Any reason why the old numbers do not look good anymore?

File: src/shapes/text.class.js [hunk 4]
```diff
@@ -737,12 +737,16 @@
...
     getCompleteStyleDeclaration: function(lineIndex, charIndex) {
       var style = this._getStyleDeclaration(lineIndex, charIndex) || { },
-          styleObject = { }, prop;
-      for (var i = 0; i < this._styleProperties.length; i++) {
-        prop = this._styleProperties[i];
-        styleObject[prop] = typeof style[prop] === 'undefined' ? this[prop] : style[prop];
+          decl = { }, slate = this._styleProperties;
+      for (var i = slate.length - 1; i >= 0; --i) {
+        var prop = slate[i];
+        if (style[prop] != null) {
```
Comment (asturur) [inline src/shapes/text.class.js]: Why the reversed for loop and why changing an undefined check with a null?

File: src/shapes/text.class.js (hunk 3 above)
Comment (asturur) [inline src/shapes/text.class.js]: the source you documented:
image

has references to behaviours of other apps.

File: src/shapes/text.class.js [hunk 5]
```diff
@@ -950,8 +954,8 @@
 
       var line = this._textLines[lineIndex], maxHeight = 0;
       for (var i = 0, len = line.length; i < len; i++) {
-        maxHeight = Math.max(this.getHeightOfChar(lineIndex, i), maxHeight);
-        // TODO: Take 'deltaY' into account to avoid overlaps (may require 'allowOverlaps' property)
+        var box = this._getGraphemeBox(null, lineIndex, i);
+        maxHeight = Math.max(maxHeight, box.height);
```
Comment (asturur) [inline src/shapes/text.class.js]: i would not change getHeightOfLine, i know devs will be able to set values that overlap, but i do not want to change basic behaviours for a new feature that has a marginal impact on overall library usage.

File: src/shapes/text.class.js (hunk 4 above)
Comment (denim2x) [inline src/shapes/text.class.js]: well, about that... when order doesn't matter, a reversed loop is usually *faster* in JS;
as for `null`, well - it's shorter for one - but it's also *equivalent* (i.e. `null == undefined`)

File: src/shapes/text.class.js (hunk 5 above)
Comment (denim2x) [inline src/shapes/text.class.js]: well then I'll have to come up with a different method - say `getExtentOfLine`

**Comment by asturur** [comment]
@denim2x 
I m stating clearly as much as possible.
I do not want to change any measurement procedure in fabricjs to introduce this feature.
I simply cannot support it.

From a wrong consideration now may come an handfull of bugs that then i have to fix or revert changes.

I do not want to change any code that is unrelated to the functionality for some supposed performance improvement that is not backed by solid benchmarks.

The functionality is as good as it is if we fix the cursor rendering.

Unless the owner of the bounty comes with some different requirements and in that case i have to weight in my decisions.

**Pull request closed without merging**